### PR TITLE
Adding script for recovering a cluster.yaml using kubectl.

### PR DESCRIPTION
**Pull request by mattmattox**
Importing https://gist.github.com/mattmattox/d32b3fea4820075c08c6cc2f6d736702

### CODE REVIEW

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+echo "Checking that yq v3.x is installed"
+if ! `yq -V | grep 'yq version 3'`
+then
+  echo "Please download yq v3.x from https://github.com/mikefarah/yq/releases/tag/3.4.1 and install it"
+  exit 1
+fi
+
+echo "Building cluster.yml..."
+echo "Working on Nodes..."
+echo 'nodes:' > cluster.yml
```
Comment (ansilh) [inline how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:25]: Add a guardrail to exit if `cluster.yml` exists.

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+if [[ ! -f kube_config_cluster.yml ]]
+then
+  echo "The kube_config_cluster.yml is missing"
+  exit 1
+fi
+
+echo "Checking that jq is installed"
+if ! jq -V
```
Comment (ansilh) [inline how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:10]: We can write this as `if [ -x jq ]`
```      
 -x file 
     True if file exists and is executable.
```

**Review by ansilh — CHANGES_REQUESTED** [review]

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh (hunk 1 above)
Comment (mattmattox) [reply how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:31]: Fixed

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh (hunk 2 above)
Comment (mattmattox) [reply how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh]: Fixed

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+echo "Checking that jq is installed"
+if [ -x jq ]
+then
+  echo "Please download jq from https://github.com/stedolan/jq/releases/tag/jq-1.6 and install it"
+  exti 1
+fi
+
+echo "Checking that yq v3.x is installed"
+if ! `yq -V | grep 'yq version 3'`
```
Comment (ansilh) [inline how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:23]: The output of the command will be considered as another command and the script tries to execute it and fails.
Need to rewrite like below

```
if [  -z  "`yq -V |grep "yq version 3"`" ]
```

```
+ [[ ! -f kube_config_cluster.yml ]]
+ [[ -f cluster.yml ]]
+ echo 'Checking that jq is installed'
Checking that jq is installed
+ '[' -x jq ']'
+ echo 'Checking that yq v3.x is installed'
Checking that yq v3.x is installed
++ grep 'yq version 3'
++ yq -V
+ yq version 3.4.1
Error: unknown command "version" for "yq"
Run 'yq --help' for usage.
+ echo 'Please download yq v3.x from https://github.com/mikefarah/yq/releases/tag/3.4.1 and install it'
Please download yq v3.x from https://github.com/mikefarah/yq/releases/tag/3.4.1 and install it
+ exit 1
```

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+then
+  echo "Please download yq v3.x from https://github.com/mikefarah/yq/releases/tag/3.4.1 and install it"
+  exit 1
+fi
+
+echo "Building cluster.yml..."
+echo "Working on Nodes..."
+echo 'nodes:' > cluster.yml
+kubectl --kubeconfig kube_config_cluster.yml -n kube-system get configmap full-cluster-state -o json | jq -r .data.\"full-cluster-state\" | jq -r .desiredState.rkeConfig.nodes | yq r -P - | sed 's/^/  /' | \
```
Comment (ansilh) [inline how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:32]: Please add a check for `kubectl` 
if not present, the script continues but trows below error.
```
line 32: kubectl: command not found
```

**Review by ansilh — CHANGES_REQUESTED** [review]
Need some minor modifications

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh (hunk 3 above)
Comment (mattmattox) [reply how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh]: Fixed

File: how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh (hunk 4 above)
Comment (mattmattox) [reply how-to-retrieve-cluster-yaml-from-custom-cluster/cluster-yaml-recovery.sh:40]: Fixed

**Review by ansilh — CHANGES_REQUESTED** [review]
Modified the script and its able to get the cluster.yml file 

But `rke up` using retrieved cluster.yml fails as it tries to pull image from public repo instead of private.
```
INFO[0000] Checking if container [cert-deployer] is running on host [192.168.110.171], try #1
INFO[0000] Pulling image [rancher/rke-tools:v0.1.69] on host [192.168.110.171], try #1
```
`rke up` using the original cluster.yml file work fine.
```
INFO[0000] Checking if container [cert-deployer] is running on host [192.168.110.171], try #1
INFO[0000] Image [registry.ansil.io/proxy/rancher/rke-tools:v0.1.65] exists on host [192.168.110.171]
```
Looks like we need the private registry stanza to do `rke up` in airgap.

Also, the image versions of rke-tools in both cases are different.

**Comment by mattmattox** [comment]
@ansilh I ended up rewriting the script to just pull the whole json file from the configmap then do a number of finds and replaces to fix and name changes using https://github.com/rancher/rke/blob/master/types/rke_types.go as a guide.

**Comment by ansilh** [comment]
Hi @mattmattox , 
`rke up` was  failing due to missing parameters and found that there were still parameters in camelcase 
I rewrote the build function as below.

The logic I've applied here is to replace `camelCase` characters before first occurrence of `:` with `_[a-z]`

Eg:- 

`kubeApi` -> `kube_api`
`kubeController` -> `kube_controller`

```
build() {
 echo "Building cluster.yml..."
 IFS=''
 while read line
 do
  word=$(echo "$line" |awk -F ":" '{print $1}')
  new_word=$(echo "$line" |awk -F ":" '{print $1}' | perl -pe 's/([a-z0-9])([A-Z])/$1_\L$2/g')
  echo "$line"| sed "s/$word/$new_word/g" >>cluster.yml
 done< <(${KUBECTL_CMD} -n kube-system get configmap full-cluster-state -o json | jq -r .data.\"full-cluster-state\" | jq -r .desiredState.rkeConfig | yq r -P -)

 echo "Building cluster.rkestate..."
 ${KUBECTL_CMD} -n kube-system get configmap full-cluster-state -o json | jq -r .data.\"full-cluster-state\" | jq -r . > cluster.rkestate
}

```

Could you please test this out?

**Pull request closed without merging**